### PR TITLE
SPLICE-983: Carrying defaults through imports.

### PR DESCRIPTION
**Pull request by scottfines**
2.0 version of SPLICE-983 fix.

It turns out, Imports are really just inserts over subselects; to
support default value during the import, we can just re-write the import
SQL clauses to include a case statement which replaces null entries with
default values.

However, in order to support this case statement effectively, we also
need to make sure that we properly treat NULL strings within the parsing
stage. Per customer request, we treat null, NULL, nUll, etc. as actually
being null; we treat "null" as the string 'null'.

An IT was added to confirm overall behavior.

Conflicts:
	splice_machine/src/main/java/com/splicemachine/derby/stream/function/FileFunction.java
	splice_machine/src/main/java/com/splicemachine/derby/stream/function/StreamFileFunction.java

### CODE REVIEW

File: splice_machine/src/main/java/com/splicemachine/derby/impl/load/HdfsImport.java [hunk 1]
```diff
@@ -407,6 +420,65 @@ private static void doImport(String schemaName,
...
+            SchemaDescriptor sd = dd.getSchemaDescriptor(schemaName,lcc.getTransactionExecute(),true);
+            assert sd!=null: "Programmer error: schema is not found!";
+            TableDescriptor td = dd.getTableDescriptor(tableName,sd,lcc.getTransactionExecute());
+            assert td!=null: "Programmer error: table is not found!";
+
+            ColumnDescriptorList columnDescriptorList=td.getColumnDescriptorList();
+            if(insertColumnList!=null){
+                boolean isFirst = true;
+                for(String col:insertColumnList){
+                    if(isFirst) isFirst = false;
```
Comment (snowspeeder) [inline splice_machine/src/main/java/com/splicemachine/derby/impl/load/HdfsImport.java]: where is it using isFirst flag, bit confusing

File: splice_machine/src/main/java/com/splicemachine/derby/stream/function/AbstractFileFunction.java [hunk 2]
```diff
@@ -180,6 +188,14 @@ void checkPreference() {
         }
     }
 
+    /* ****************************************************************************************************************/
+    /*private helper methods*/
+    @SuppressWarnings("SimplifiableIfStatement") //the logic is clearer this way, without a performance penalty
+    private boolean shouldBeNull(String value,boolean wasQuoted){
```
Comment (snowspeeder) [inline splice_machine/src/main/java/com/splicemachine/derby/stream/function/AbstractFileFunction.java]: add static?

**Comment by genedavis** [comment]
+1

**Comment by jyuanca** [comment]
+1

**Comment by jfilali** [comment]
+1

File: splice_machine/src/main/java/com/splicemachine/derby/stream/function/QuoteTrackingTokenizer.java [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+        // read a line (ignoring empty lines/comments if necessary)
+        String line;
+        do{
+            line=readLine();
+            if(line==null){
+                return false; // EOF
+            }
+        }
+        while(ignoreEmptyLines && line.length()==0 || (commentMatcher!=null && commentMatcher.isComment(line)));
```
Comment (snowspeeder) [inline splice_machine/src/main/java/com/splicemachine/derby/stream/function/QuoteTrackingTokenizer.java]: is it possible that line may contain other white spaces, or it needs trim before length check?

File: splice_machine/src/main/java/com/splicemachine/derby/stream/function/QuoteTrackingTokenizer.java [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+                        appendSpaces(currentColumn,potentialSpaces);
+                    }
+                    columns.add(currentColumn.length()>0?currentColumn.toString():null); // "" -> null
+                    potentialSpaces=0;
+                    currentColumn.setLength(0);
+                    if(recordQuotes){
+                        quotedColumns.append(wasQuoted);
+                    }
+                    wasQuoted = false;
+                }else if(c==SPACE){
```
Comment (snowspeeder) [inline splice_machine/src/main/java/com/splicemachine/derby/stream/function/QuoteTrackingTokenizer.java]: how about other whitespace chars, \t,\r etc

File: splice_machine/src/main/java/com/splicemachine/derby/stream/function/QuoteTrackingTokenizer.java [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+                    if(!surroundingSpacesNeedQuotes || currentColumn.length()>0){
+                        appendSpaces(currentColumn,potentialSpaces);
+                    }
+
+                    potentialSpaces=0;
+                    currentColumn.append(c);
+                }
+
+            }else{
+				/*
```
Comment (snowspeeder) [inline splice_machine/src/main/java/com/splicemachine/derby/stream/function/QuoteTrackingTokenizer.java]: nice to fix mis-alignments in comments

**Pull request merged**